coordinator-common/src/test/java/org/apache/kafka/coordinator/common/runtime/HdrHistogramTest.java

```diff
@@ -172,4 +177,39 @@ public void testHistogramDataReset() {
         assertEquals(numEventsInFirstCycle, hdrHistogram.count(now + maxSnapshotAgeMs));
         assertEquals(numEventsInSecondCycle, hdrHistogram.count(now + 1 + maxSnapshotAgeMs));
     }
+
+    @Test
+    public void testLatestHistogramRace() throws InterruptedException, ExecutionException {
+        long maxSnapshotAgeMs = 10L;
+        long now = System.currentTimeMillis();
+        HdrHistogram hdrHistogram = new HdrHistogram(maxSnapshotAgeMs, MAX_VALUE, 1);
+        ExecutorService countExecutor = Executors.newFixedThreadPool(2);
+        for (int i = 1; i < 10000; i++) {
+            int numEvents = 2;
+            for (int j = 0; j < numEvents; j++) {
+                hdrHistogram.record(i);
+            }
+            final long moreThanMaxAge = now + maxSnapshotAgeMs + 1;
+            now = moreThanMaxAge;
+            CountDownLatch latch = new CountDownLatch(1);
+            Callable<Long> countTask = () -> {
+                try {
+                    assertTrue(latch.await(500, TimeUnit.MILLISECONDS));
```

Comment:
if we want both threads to arrive here before proceeding, i think we can use CyclicBarrier instead

Reply:
Ah, thanks for raising that, I actually considered this and there are two reasons why I didn't go with a barrier:
1. The latch is more familiar to people - e.g. in the AK codebase I can see >100 matches in >30 files for `CountDownLatch` and <10 matches in only 2 files for `CyclicBarrier`.
2. On my local machine the latch is faster, resulting in a local run of the test between 400 and 500 ms, while the barrier results in a local run of the test between 500 and 700 ms, regardless if the barrier is reused and reset or not. This matches my previous experience and can be explained by the fact that the barrier locks manually while the latch uses AQS (the main building block of the various locking and synchronization utilities the JDK provides) directly.
    - Note that we can also use a `Phaser` here, and while it's faster than the barrier, it's even less well-known than it, it requires 2 API calls if we want to have a timed wait, and is still not really as fast as the latch.

Just to clarify, what we are doing here is a canonical usage of a `CountDownLatch` - signalling the worker threads from the external test thread.

Reply:
my understanding is that the main thread may count down the latch before the threads reach latch.await() which then loses its purpose. this could happen right?

regardless, i think your explanation makes sense. thanks!